### PR TITLE
Create vspheredatacenterconfig reconciler

### PR DESCRIPTION
**Pull request by jiayiwang7**
… that validates vcenter config and update its status

*Issue #, if available:*

#870 

*Description of changes:*

API changes
- Add `SpecValid` to `VSphereDatacenterConfigStatus`

Doc changes
- Add VSphereDatacenterConfigStatus session to cluster full lifecycle API design

Code changes
- Refactor vsphere `setup` method
- export vsphere validator and defaulter struct
- initialize validator and defaulter during controller setupmanager time
- fetch `vsphere-credentials` secret from cluster and use it for Govc calls
- validate vcenterconfig and update its status within reconciler

*Follow up*
- Make govc executable independent from using Env Vars
- Add `FailureReason` and `FailureMessage` to `VSphereDatacenterConfigStatus`
- Unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by g-gaston — COMMENTED** [review]
This is great! 🔝 
Just some minor comments, mostly questions 🚀

File: controllers/controllers/vsphere_datacenter_controller.go [hunk 1]
```diff
@@ -25,15 +36,113 @@ func NewVSphereDatacenterReconciler(client client.Client, log logr.Logger, schem
...
 
 // SetupWithManager sets up the controller with the Manager.
-func (r *VSphereDatacenterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *VSphereDatacenterReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	deps, err := dependencies.NewFactory().WithGovc().Build(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.validator = vsphere.NewValidator(deps.Govc, &networkutils.DefaultNetClient{})
```
[the remark this answers is not in the document]
Comment (jiayiwang7) [reply controllers/controllers/vsphere_datacenter_controller.go:45]: what's the downside of adding the validator/defaulter in `SetupWithManager`? I saw CAPI does similar things

File: controllers/controllers/vsphere_datacenter_controller.go [hunk 2]
```diff
@@ -25,15 +36,113 @@ func NewVSphereDatacenterReconciler(client client.Client, log logr.Logger, schem
...
+
+	if err := os.Setenv(vsphere.EksavSpherePasswordKey, string(vspherePassword)); err != nil {
+		return fmt.Errorf("failed setting env %s: %v", vsphere.EksavSpherePasswordKey, err)
+	}
+
+	if err := vsphere.SetupEnvVars(vsphereDatacenter); err != nil {
+		return fmt.Errorf("failed setting env vars: %v", err)
+	}
+
+	if err := r.defaulter.SetDefaultsForDatacenterConfig(ctx, vsphereDatacenter); err != nil {
```
[the remark this answers is not in the document]
Comment (jiayiwang7) [reply controllers/controllers/vsphere_datacenter_controller.go:139]: many of the defaults checks requires `govc`, are you suggesting passing it to the API as arg? or only move the ones without `govc` dep

[the remark this answers is not in the document]
Comment (g-gaston) [reply controllers/controllers/vsphere_datacenter_controller.go:139]: Only the ones without `govc` dep :)

File: controllers/controllers/vsphere_datacenter_controller.go [hunk 3]
```diff
@@ -25,15 +36,113 @@ func NewVSphereDatacenterReconciler(client client.Client, log logr.Logger, schem
...
+		return ctrl.Result{}, err
+	}
+	// Determine if VsphereDatacenterConfig is valid
+	if err := r.validator.ValidateVCenterConfig(ctx, vsphereDatacenter); err != nil {
+		log.Error(err, "Failed to validate VsphereDatacenterConfig")
+		vsphereDatacenter.Status.SpecValid = false
+		return ctrl.Result{}, err
+	}
+
+	vsphereDatacenter.Status.SpecValid = true
```
[the remark this answers is not in the document]
Comment (jiayiwang7) [reply controllers/controllers/vsphere_datacenter_controller.go:101]: will handle in follow up PR

File: controllers/controllers/vsphere_datacenter_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jiayiwang7) [reply controllers/controllers/vsphere_datacenter_controller.go:139]: will handle in followup PR

**Review by mdsgabriel — APPROVED** [review]

**Pull request merged**